### PR TITLE
OCT-1803: Adds migration to change uq score for users

### PR DESCRIPTION
**Pull request by adam-gf**
## Description
Adds migration to change uq score for users

## Definition of Done

1. [ ] Acceptance criteria are met.
2. [ ] PR is manually tested before the merge by developer(s).
    - [ ] Happy path is manually checked.
3. [ ] PR is manually tested by QA when their assistance is required (1).
    - [ ] Octant Areas & Test Cases are checked for impact and updated if required (2).
4. [ ] Unit tests are added unless there is a reason to omit them.
5. [ ] Automated tests are added when required.
6. [ ] The code is merged.
7. [ ] Tech documentation is added / updated, reviewed and approved (including mandatory approval by a code owner, should such exist for changed files).
    - [ ] BE: Swagger documentation is updated.
8. [ ] When required by QA:
    - [ ] Deployed to the relevant environment.
    - [ ] Passed system tests.

---

(1) Developer(s) in coordination with QA decide whether it's required. For small tickets introducing small changes QA assistance is most probably not required.

(2) [Octant Areas & Test Cases](https://docs.google.com/spreadsheets/d/1cRe6dxuKJV3a4ZskAwWEPvrFkQm6rEfyUCYwLTYw_Cc).

### CODE REVIEW

File: backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+    "0x848e313d4b25bc0b48cafdb6a72391e892e6a247",
+    "0x0025ab2d69f6c2c3ffac32ab6a16e18c807518b8",
+    "0x2efe744ecc4f6bd55538da57d09dae895c95b223",
+    "0xbc6d82d8d6632938394905bb0217ad9c673015d1",
+    "0xe1555c6ee61366a3f90135dc704acd25c3247aca",
+    "0x2f51e78ff8aec6a941c4ceeeb26b4a1f03737c50",
+]
+
+
+upgrade()
```
Comment (kgarbacinski) [inline backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py:519]: why is it here?

File: backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+Revision ID: 5d3bf271b78d
+Revises: 19f6b4f5ce2e
+Create Date: 2024-07-19 14:11:00.850135
+
+"""
+from alembic import op
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
```
Comment (kgarbacinski) [inline backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py:12]: to remove

File: backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+            SELECT id FROM users WHERE address IN (
+                {", ".join([f"'{address}'" for address in user_addresses])}
+            )       
+        );
+    """
+    op.execute(query)
+
+
+def downgrade():
+    # Revert the changes made in the upgrade
```
Comment (kgarbacinski) [inline backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py:35]: comment to remove

File: backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+depends_on = None
+
+
+def upgrade():
+    # Waitlist is a list of stings that represent user addresses
+    # For all users in the waitlist, if they have a UQ score, update it to 1.0 only for epoch 5
+    query = f"""
+        UPDATE uniqueness_quotients
+        SET score = '1.0'
+        WHERE epoch = 5 AND user_id IN (
```
Comment (kgarbacinski) [inline backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py:25]: I think we need it for E4, not E5

Comment (kgarbacinski) [inline backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py:25]: you can also consider adding another predicate: ```WHERE score = 0.2```

File: backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+    """
+    op.execute(query)
+
+
+def downgrade():
+    query = f"""
+        UPDATE uniqueness_quotients
+        SET score = '0.2'
+        WHERE score = '1.0' AND epoch = 4 AND user_id IN (
+            SELECT id FROM users WHERE address IN (
```
Comment (paulperegud) [inline backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py:38]: Strictly speaking this downgrade doesn't reverse the upgrade. Not sure if we should care about it ATM.

**Review by paulperegud — DISMISSED** [review]

File: backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py [hunk 6]
```diff
@@ -0,0 +1,514 @@
+"""update uq score for waitlist
```
Comment (paulperegud) [inline backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py:1]: Why "waitlist"? Isn't it a guest list?

Comment (aziolek) [reply backend/migrations/versions/5d3bf271b78d_update_uq_score_for_waitlist.py:1]: Either `whitelist` or `guestlist`.

**Review by aziolek — CHANGES_REQUESTED** [review]
This waits for https://github.com/golemfoundation/octant/pull/350.

**Comment by paulperegud** [comment]
@adam-gf please also recheck if addresses should be normalized before comparison with addresses in our DB. AFAIR we store all addresses in normalized with `to_checksum_address(...)`.

File: backend/migrations/versions/5d3bf271b78d_update_uq_score_for_guestlist.py [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+                {", ".join([f"'{address}'" for address in user_addresses])}
+            )       
+        );
+    """
+    op.execute(query)
+
+
+def downgrade():
+    query = f"""
+        UPDATE uniqueness_quotients
```
Comment (kgarbacinski) [inline backend/migrations/versions/5d3bf271b78d_update_uq_score_for_guestlist.py:36]: the possible issue here is that we'll downgrade every person from the guestlist (not only ppl from the upgrade) but it's not very likely to execute this so not a blocker

**Review by kgarbacinski — APPROVED** [review]

**Review by paulperegud — APPROVED** [review]

File: backend/migrations/versions/5d3bf271b78d_update_uq_score_for_guestlist.py [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+        WHERE score = '1.0' AND epoch = 4 AND user_id IN (
+            SELECT id FROM users WHERE address IN (
+                {", ".join([f"'{address}'" for address in user_addresses])}
+            )       
+        );
+    """
+    op.execute(query)
+
+
+raw_addresses = [
```
[the remark this answers is not in the document]
Comment (adam-gf) [reply backend/migrations/versions/5d3bf271b78d_update_uq_score_for_guestlist.py:47]: Only for those who have UQ score  set in db, so only those who voted.

**Comment by aziolek** [comment]
https://github.com/golemfoundation/octant/pull/350 is merged.

**Review by 0xartem — APPROVED** [review]

**Review by aziolek — APPROVED** [review]

**Pull request merged**